### PR TITLE
Improve performance of graph automorphism via faster refinement and r…

### PR DESCRIPTION
**Pull request by SebastianGitt**
…emoval of ``change_base()``

- uses an optimized colour refinement algorithm
- more efficiently handles graphs with disjoint components
- fixes issue where automorphisms could be missed if coset representatives could belong to multiple left transversals
- supports canonical labelling
- faster graph comparison using refinement trace
- no longer relies on ``change_base()`` to perform pruning by automorphism
- uses more efficient data structures
- now returns mapping to original graph labelling
- improves test coverage of edge cases

### CODE REVIEW

File: dwave/experimental/automorphism/automorphism_generation.py [hunk 1]
```diff
@@ -126,163 +191,449 @@ def _test_composability(self, g: NDArray[np.intp]) -> tuple[int, NDArray[np.intp
...
+        trace: NDArray[np.integer],
+        active_vertices: list[list[int]],
+        colour_degree: list[int],
+        min_degree: int,
+        max_degree: int,
+        refine_stack: list[int],
+        in_refine_stack: bytearray,
+        next_colour: list[int],
+    ) -> None:
+        """A subroutine of ``_refine`` that splits a colour class into subcells
```
Comment (SebastianGitt) [inline dwave/experimental/automorphism/automorphism_generation.py]: forgot the brackets in ``refine()``

File: tests/test_automorphism.py [hunk 2]
```diff
@@ -267,16 +283,302 @@ def test_schreier_null_zero(self):
...
+
+        self.assertEqual(len(result.vertex_orbits), expected_num_vertex_orbits)
+
+
+    def test_nauty(self):
+        """Check the number of automorphisms, vertex orbits, and edge orbits of the graph from
+        the example from McKay & Piperno (2014), "Practical Graph Isomorphism II"."""
+
+        graph = nx.Graph()
+        graph.add_edges_from([
```
Comment (SebastianGitt) [inline tests/test_automorphism.py]: fewer lines?

File: tests/test_automorphism.py [hunk 3]
```diff
@@ -267,16 +283,302 @@ def test_schreier_null_zero(self):
...
+        expected_node_to_index =  {0: 0, 1: 1, 2: 2, 3: 3, 4: 4, 5: 5, 6: 6, 7: 7, 8: 8}
+        self.assertEqual(result.index_to_node, expected_node_to_index)
+
+    def test_node_mappings_nauty_shifted(self):
+        """Test the `index_to_node` and `node_to_index` properties for a graph
+        with node labels equal to a fixed shift from its node indices."""
+
+        graph = nx.Graph()
+        graph.add_edges_from([
+            (3, 4), (3, 6),
```
Comment (SebastianGitt) [inline tests/test_automorphism.py]: condense

File: dwave/experimental/automorphism/automorphism_generation.py [hunk 4]
```diff
@@ -86,26 +135,42 @@ def num_automorphisms(self) -> int:
...
-        """Edge orbits induced by the coset representatives in u_vector."""
-        return edge_orbits(self._u_vector, self._graph_edges)
+    def vertex_orbits_original_labels(self) -> list[list[Hashable]]:
+        """Vertex orbits induced by the coset representatives in u_vector and returned
+        with the original node labels."""
+        mapped_orbits = []
+        for orbit in vertex_orbits(self._u_vector, self._nodes):
+            mapped_orbit = []
+            for index in orbit:
+                mapped_orbit.append(self._index_to_node[index])
```
Comment (randomir) [inline dwave/experimental/automorphism/automorphism_generation.py]:
```suggestion
            mapped_orbit = [self._index_to_node[index] for index in orbit]
```

File: dwave/experimental/automorphism/automorphism_generation.py [hunk 5]
```diff
@@ -86,26 +135,42 @@ def num_automorphisms(self) -> int:
...
-        return self._rng.sample(generators, self._num_samples)
+    @property
+    def edge_orbits_original_labels(self) -> list[list[Hashable]]:
+        """Edge orbits induced by the coset representatives in u_vector and returned
+        with the original node labels."""
+        mapped_orbits = []
+        for orbit in edge_orbits(self._u_vector, self._graph_edges):
+            mapped_orbit = []
+            for (u, v) in orbit:
+                mapped_orbit.append((self._index_to_node[u], self._index_to_node[v]))
```
Comment (randomir) [inline dwave/experimental/automorphism/automorphism_generation.py]: Using list comprehension here (as also suggested above) should be faster (unless dominated by the node lookup) -- around 30% on my laptop.

```suggestion
            mapped_orbit = [(self._index_to_node[u], self._index_to_node[v]) for (u, v) in orbit]
```

File: dwave/experimental/automorphism/automorphism_generation.py [hunk 6]
```diff
@@ -86,26 +135,42 @@ def num_automorphisms(self) -> int:
...
+            mapped_orbits.append(sorted(mapped_orbit))
+
+        return sorted(mapped_orbits)
```
Comment (randomir) [inline dwave/experimental/automorphism/automorphism_generation.py]: Inplace sort makes more sense as this function owns the source list already.

**Review by randomir — APPROVED** [review]
Looks good, just a few minor suggestions.

I'll defer the algorithm validation to @jackraymond.

File: docs/index.rst
Comment (jackraymond) [inline docs/index.rst:1]: Why is github marking these files as changed? Shouldn't be touched, maybe a technicality?

File: dwave/experimental/automorphism/automorphism_generation.py [hunk 7]
```diff
@@ -126,163 +178,459 @@ def _test_composability(self, g: NDArray[np.intp]) -> tuple[int, NDArray[np.intp
...
+        ignoring the automorphisms that do not respect the current partition structure. 
 
         Args:
-            partition: Partition represented as a list of vertex index sets.
+            partition: The current partition structure, represented as a list of
+                sets of vertices ordered by color.
+            trace: The number of vertices belonging to each color, ordered by color.
+            color: A map from each vertex to its color.
+            num_colors: The number of unique colors, equivalent to the number
+            of cells in the partition.
```
Comment (jackraymond) [inline dwave/experimental/automorphism/automorphism_generation.py]: indent.
```suggestion
                of cells in the partition.
```

**Review by jackraymond — APPROVED** [review]
Changes look good to me.

There is some issue with the file-comparison here. It marks almost all files as new, but I know only .

I've functionally tested this in the context of AutomorphismComposite with O(10) embeddings each of size 128 (periodic embedded cubic lattices) in Advantage2 and Advantage systems, seeing a significant jump in performance.

File: dwave/experimental/automorphism/automorphism_generation.py [hunk 8]
```diff
@@ -126,163 +178,435 @@ def _test_composability(self, g: NDArray[np.intp]) -> tuple[int, NDArray[np.intp
...
+        the refinement stack.
+
+        The process continues until no color class can be further refined, yielding an
+        equitable coloring.
+
+        If a vertex was individualized prior to this refinement step, only the
+        color class containing that vertex needs to be placed on the refinement
+        stack initially, since only colors adjacent to that color can be affected.
+
+        For performance reasons, `num_colors` is passed as a single-element list to
```
Comment (thisac) [inline dwave/experimental/automorphism/automorphism_generation.py]:
```suggestion
        For performance reasons, ``num_colors`` is passed as a single-element list to
```

File: dwave/experimental/automorphism/automorphism_generation.py [hunk 9]
```diff
@@ -126,163 +178,435 @@ def _test_composability(self, g: NDArray[np.intp]) -> tuple[int, NDArray[np.intp
...
+        `_split_up_color()` so that updates to the number of colors persist across
```
Comment (thisac) [inline dwave/experimental/automorphism/automorphism_generation.py]: Generally, I would try not to refer to hidden methods in docstrings, but this is also in a hidden method so not that important here. Still, I think it's better to update this reference.

File: dwave/experimental/automorphism/automorphism_generation.py [hunk 10]
```diff
@@ -126,163 +178,435 @@ def _test_composability(self, g: NDArray[np.intp]) -> tuple[int, NDArray[np.intp
...
+        trace: NDArray[np.integer],
+        active_vertices: list[list[int]],
+        color_degree: list[int],
+        min_degree: int,
+        max_degree: int,
+        refine_stack: list[int],
+        in_refine_stack: bytearray,
+        num_colors: list[int],
+    ) -> None:
+        """A subroutine of ``_refine()`` that splits a color class into subcells
```
Comment (thisac) [inline dwave/experimental/automorphism/automorphism_generation.py]: Same here w.r.t. hidden method reference.

File: dwave/experimental/automorphism/automorphism_generation.py [hunk 11]
```diff
@@ -126,163 +178,435 @@ def _test_composability(self, g: NDArray[np.intp]) -> tuple[int, NDArray[np.intp
...
+                trace[new_color] += 1
+                color[v] = new_color
+
+    def _canon(
+            self,
+            partition: list[set[int]],
+            trace: NDArray[np.integer],
+            color: NDArray[np.integer],
+            num_colors: int,
+            individualized_vertex: Optional[int] = None,
```
Comment (thisac) [inline dwave/experimental/automorphism/automorphism_generation.py]:
```suggestion
        self,
        partition: list[set[int]],
        trace: NDArray[np.integer],
        color: NDArray[np.integer],
        num_colors: int,
        individualized_vertex: Optional[int] = None,
```

Comment (thisac) [reply dwave/experimental/automorphism/automorphism_generation.py]: We can also start using `int | None` instead of `Optional[int]` if we'd like. IMO slightly cleaner, and I think we're leaning towards that elsewhere.

File: docs/index.rst
Comment (SebastianGitt) [reply docs/index.rst]: This should be fixed now.

**Review by thisac — APPROVED** [review]
